src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1023 @@
+Only one option set is allowed and only its existence, not its value is tested.
+In addition, the condition may be negated (option or option_set must
+not exist) by prepending it with '!'.
+
+=item option_name 
+
+the name of an option that will be retrieved from the configuration.
+
+=item option_set 
+
+the name of an option set where the option is located in (for example 'dpnsHost:dpm'
+means C<dpnsHost> option of 'dpm' option set. An option set is a sub-hash in the configuration
+hash. C<GLOBAL> is a special value for 'option_set' indicating that the option is a global option,
+instead of belonging to a specific option set (global options are at the top level of the option
+hash).
+
+=item line_fmt 
+
+defines the format used to represent the key/value pair. 3 formats are
+supported (see LINE_FORMAT_xxx constants below):
+
+=over
+
+=item 
+
+a sh shell environment variable definition (export VAR=val)
+
+=item 
+
+a sh shell variable definition (VAR=val)
+
+=item 
+
+a 'keyword value' line, as used by Xrootd or Apache config files.
+
+=item 
+
+a 'setenv keyword value' line, as used by Xrootd config files mainly.
+
+=item 
+
+a 'set keyword value' line, as used by Xrootd config files mainly.
+
+=back
+
+Inline comments are not supported in 'keyword value' family of formats.
+
+=item value_fmt 
+
+used to indicate how to interpret the configuration value. It is used mainly for
+boolean values, list and hashes. See LINE_VALUE_xxx constants below for the possible values.
+
+=item value_fmt 
+
+used to indicate how to interpret the configuration value. It is used mainly for
+boolean values, list and hashes. See LINE_VALUE_xxx constants below for the possible values.
+
+=back 
+
+For an example of rules, look at ncm-dpmlfc or ncm-xrootd source code in
+configuration-modules-grid repository.
+
+
+=head2 Public methods
+
+=over
+
+=item new
+
+This is the constructor. It mainly executes the FileEditor constructor and supports the
+same arguments.
+
+==cut
+
+sub new
+{
+    my $class = shift;
+    my $self = $class->SUPER::new (@_);
+    return $self;
+}
+
+
+=pod
+
+=item updateFile
+
+Update configuration file contents,  applying configuration rules.
+
+Arguments :
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+Return value
+    always 0
+    
+=cut
+
+sub updateFile {
+  my $function_name = "updateConfigFile";
+  my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+  unless ( $config_rules ) {
+    *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
```

Comment:
is this the code that fails and that you pm'ed me about. it's pretty normal that this fails without defininig a log instance. you have to use `$self->error()` if you want this to work without defining log instance